src/coreclr/vm/siginfo.cpp

```diff
@@ -3782,10 +3785,40 @@ MetaSig::CompareElementType(
                 }
             }
         }
-        else
+        else if (state->IgnoreCustomModifiers)
         {
-            return FALSE; // types must be the same
+            mdToken tk;
+            bool consumedCustomModifier = false;
+            switch (Type1)
+            {
+            case ELEMENT_TYPE_CMOD_REQD:
+            case ELEMENT_TYPE_CMOD_OPT:
+                IfFailThrow(CorSigUncompressToken_EndPtr(pSig1, pEndSig1, &tk));
+                state->CustomModifierCount1++;
```

Comment:
Here is the full algorithm from C# compiler to pick winners between ambiguous overloads: https://github.com/dotnet/roslyn/blob/826e1770eedb6d8d44a054c9e3ea6569e7ec5970/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs#L1708-L2135. We are taking 5 lines from the end of this 400 line algorithm. I am not sure whether it makes sense. There are number of other higher priority checks before the custom modifier count tie breaker that we can use too, but I assume that we are not going to implement those.

Reply:
Consider this case:

```
[UnsafeAccessor(UnsafeAccessorKind.Method)]
void M(C @this, delegate* unmanaged[Stdcall, MemberFunction]<void> pfn) { }

unsafe class C
{
    void M(delegate* unmanaged<void> pfn) { }

    void M(delegate* unmanaged[Stdcall, MemberFunction]<void> pfn) { }
}
```

If we ignore modops/modreqs and pick the overload with fewer of them as the winner, my understanding is that we will bind to `void M(delegate* unmanaged<void> pfn) { }` in this case. It is non-sensical behavior. The author of the code clearly went extra length to bind to the other overload.

> I don't think duplicating all of Roslyn's overload logic is valuable.

Agree.

> If the counts don't match at present, it means there is an ambiguity and I will be throwing an exception.

It is fine behavior. We do not need to try to "fix" this. It is not the only source of corner-case ambiguous matches.

If we wanted to do something here, it may make sense to prefer exact matches of modops/modreqs over inexact matches modops/modreqs: If there is exact match of modops/modreqs, pick it as the winner. If there is match with ignored modops/modreqs and there is no ambiguity, pick it as the winner. Otherwise, throw ambiguous match exception otherwise.

Reply:
The updated logic is as follows:

1) Compare ignoring custom modifiers. If there is a single match, return.
2) Compare requiriing precise custom modifier match. If there is a single match, return.
3) Otherwise, throw `AmbiguousImplementationException`.

Reply:
Updated exception to be `AmbiguousMatchException`.

Reply:
My previously and just recently merged PR #84512  adds detailed information on what the ambiguity is when throwing `AmbigousMatchException`. Do you want me to suggest code for that here?

Reply:
I think the current behavior is what we have spec'd out and is appropriate given the scope of this feature.